File: leetcode/editor/cn/model/tree_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# 类定义过程中使用注解引用类本身
from __future__ import annotations

from typing import List, Optional


class TreeNode:

    def __init__(self, val: int = 0,
                 left: TreeNode = None,
                 right: TreeNode = None):
        self.val = val
        self.left = left
        self.right = right

    # 类方法
    @classmethod
    def decode(cls, nodes: List[int]) -> Optional[TreeNode]:
        if len(nodes) == 0:
            return None
        root = cls(nodes.pop(0))
        level = [root]
        while len(level) > 0:
            r = level.pop(0)
            if not r:
                continue
            left_filled: bool = False
            for i in range(min(2, len(nodes))):
                n = nodes.pop(0)
                # 三目运算符
                node = cls(n) if n else None
                if not left_filled:
                    r.left = node
                    left_filled = True
                else:
                    r.right = node
                level.append(node)
        return root

File: leetcode/editor/cn/model/TreeNode.java
package model;

import com.alibaba.fastjson.JSON;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class TreeNode {

    public int val;
    public TreeNode left;
    public TreeNode right;

    public static TreeNode decode(String nodesStr) {
        List<Integer> nodes = JSON.parseArray(nodesStr, Integer.class);
        if (nodes.isEmpty()) {
            return null;
        }
        TreeNode root = new TreeNode(nodes.remove(0));
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode r = queue.poll();
            if (r == null) {
                continue;
            }
            boolean leftFilled = false;
            for (int i = 0; i < Math.min(2, nodes.size()); i++) {
                Integer val = nodes.remove(0);
                TreeNode node = val == null ? null : new TreeNode(val);
                if (!leftFilled) {
                    r.left = node;
                    leftFilled = true;
                } else {
                    r.right = node;
                }
                queue.offer(node);
            }
        }
        return root;
    }

    public TreeNode(int x) {
        val = x;
    }

    public TreeNode(int x, TreeNode left, TreeNode right) {
        this.val = x;
        this.left = left;
        this.right = right;
    }

}
